tests/clitests/runtests.py

```diff
@@ -19,27 +24,106 @@
-            signal.alarm(timeout_time) # triger alarm in timeout_time seconds
+def getEngine(engine_exec):
+    rawname = os.environ.get('CASPERJS_ENGINE', engine_exec)
+    rawname = os.path.basename(rawname)
+    name = re.match('^[a-zA-Z]*', rawname)
+    if None == name:
+        print("Could not get engine name from %s\n" % (rawname))
+        sys.exit(1)
+    name = name.group(0).lower()
+    cmd_args = [engine_exec, '--version']
+    version = subprocess.check_output(cmd_args).strip()
+    parts = re.match('^[^0-9]*([0-9]+)\.([0-9]+)\.([^\s])', version)
+    if None == parts:
+        print("Could not get engine version from %s\n" % (version))
+        sys.exit(1)
+    return {
+        'NAME': name,
+        'VERSION': {
+            'MAJOR': parts.group(1),
+            'MINOR': parts.group(2),
+            'PATCH': parts.group(3)
+        }
+    }
+
+ENGINE = getEngine(ENGINE_EXEC)
+
+print("ENGINE %s" % ENGINE)
+
+# FIXME: slimerjs is not yet ready to be used as CLI because it is not
+# possible to pass arguments to the main script with slimerjs
+if 'slimerjs' == ENGINE['NAME']:
+    print("Skip cli tests for slimerjs")
+    sys.exit(0)
+
+# timeout handling as per https://gist.github.com/kirpit/1306188
+# Based on jcollado's solution:
+# http://stackoverflow.com/questions/1191374/subprocess-with-timeout/4825933#4825933
+# using ideas from https://gist.github.com/wkettler/9235609
+class Timeout(Exception):
+    def __init__(self, cmd, timeout, output=None, err=None):
+        self.cmd = cmd
+        self.timeout = timeout
+        self.output = output
+        self.err = err
+
+    def __str__(self):
+        return "Command '%s' timed out after %d second(s)." % \
+               (self.cmd, self.timeout)
+
+
+class Retcode(Exception):
+    def __init__(self, cmd, retcode, output=None, err=None):
+        self.cmd = cmd
+        self.retcode = retcode
+        self.output = output
+        self.err = err
+
+    def __str__(self):
+        return "Command '%s' returned non-zero exit status %d" % \
+               (self.cmd, self.returncode)
+
+
+class Command(object):
+    command = None
+    process = None
+    status = None
+    output, error = '', ''
+
+    def __init__(self, command):
+        if isinstance(command, basestring):
+            command = shlex.split(command)
+        self.command = command
+
+    def __str__(self):
+        return "'%s'" % (' '.join(self.command))
+
+    def run(self, timeout=None, **kwargs):
+        def target(**kwargs):
             try:
-                retval = f(*args)
-            except TimeoutException:
-                raise AssertionError('timeout of %ds. exhausted' % timeout_time)
-            finally:
-                signal.signal(signal.SIGALRM, old_handler)
-            signal.alarm(0)
-            return retval
-        return f2
-    return timeout_function
+                self.process = subprocess.Popen(self.command, **kwargs)
```

Comment:
I remember of performance issues using `subprocess.Popen` instead of `os.execvp`; what's the reasoning behind this change?